### DISCUSSION TITLE
Subscribe does not fire when something is published

### DISCUSSION THREAD

**Discussion by mStirner** [crossbario/autobahn-js#575] [Q&A]
I was using your example from the README, but the example does not work as expected.
The `onevent` call back is not executed.

It does only get called when the event is published from a other session. E.g.:

```js
try {
    // for Node.js
    var autobahn = require('autobahn');
} catch (e) {
    // for browsers (where AutobahnJS is available globally)
}

var connection1 = new autobahn.Connection({
    url: 'ws://127.0.0.1:8000/',
    realm: 'com.example.inge'
});

var connection2 = new autobahn.Connection({
    url: 'ws://127.0.0.1:8000/',
    realm: 'com.example.inge'
});

connection1.onopen = function (session) {

    // 1) subscribe to a topic
    function onevent(args) {
        console.log("Event:", args[0]);
    }
    session.subscribe('com.myapp.hello', onevent);

    // 3) register a procedure for remoting
    function add2(args) {
        return args[0] + args[1];
    }

    session.register('com.myapp.add2', add2);

};

connection2.onopen = function (session) {

    // 2) publish an event
    session.publish('com.myapp.hello', ['Hello, world!']);

    // 4) call a remote procedure
    session.call('com.myapp.add2', [2, 3]).then(
        function (res) {
            console.log("Result:", res);
        }
    );

};

connection1.open();
connection2.open();
```

The example from the readme does not work.
Is this "as it should be" or a issue with autobhan.js?

But why does `call` work in a single session, but not `publish`?

**Comment by oberstet** [comment]
When publishing an event on a session, that session is - by default - excluded from receiving its own published events even when subscribed - it will however receive events published to that topic -- published from _other_ sessions

This default behavior may be changed by publishing with option `{ exclude_me: false }`

Please see here for spec and example:

* https://wamp-proto.org/wamp_latest_ietf.html#name-publisher-exclusion
* https://github.com/crossbario/autobahn-js/blob/master/packages/autobahn/test/test_pubsub_excludeme.js

**Reply by mStirner** [reply]
> When publishing an event on a session, that session is - by default - excluded from receiving its own published events

Thanks for clearing that out, but why does this not apply to ```call``` 's?

For someone who starts with WAMP/Autobahn.js, the example is misleading, because it suggests that the event would also fire when published from the same session, exactly like `call` does.

**Comment by oberstet** [comment]
Because for PubSub, when subscribed, event payloads are received exactly with what was published, and hence if subscriber is the same as publisher, that is pointless.

For RPC, when registered, invocation payloads are received exactly with what the caller started the call with, but the _result_ (and side effects of executing the call) are not known before the call returns (later than call origin), regardless of whether the callee is identical with the caller or not. Of course the caller might as well had the callee called locally, without routing, but that's another aspect.

> the example is misleading, because it suggests that the event would also fire when published from the same session

How does the example suggest that? 

But in any case, if this is confusing, the example should explicitly discuss the aspect to save user's scratching their head and wasting time;) PRs are welcome!

Please also note that WAMP PubSub also includes even more powerful ways of restricting the set of subscribers:

https://wamp-proto.org/wamp_latest_ietf.html#name-subscriber-black-and-whitel

**Reply by mStirner** [reply]
> How does the example suggest that?

```js
   // 1) subscribe to a topic
   function onevent(args) {
      console.log("Event:", args[0]);
   }
   session.subscribe('com.myapp.hello', onevent);

   // 2) publish an event
   session.publish('com.myapp.hello', ['Hello, world!']);
```

How is this not misleading?
This is literally suggesting that when you publish a event, you immediate receive it within the same session.
I was expecting the output:

```
Event: Hello World
```

Especially when `call(...)` from the same example output something.

**Comment by oberstet** [comment]
sure! so the example should be changed to

```javascript
session.publish('com.myapp.hello', ['Hello, world!'], {}, {exclude_me: false});
```
